system_metrics_collector/src/system_metrics_collector/periodic_measurement_node.cpp

```diff
@@ -81,6 +78,10 @@ bool PeriodicMeasurementNode::setupStop()
     measurement_timer_->cancel();
     measurement_timer_.reset();
   }
+  if (publish_timer_) {
+    publish_timer_->cancel();
+    publish_timer_.reset();
+  }
```

Comment:
see my previous feedback: let's throw if `setupStop()` is called but the node is already called.

Reply:
Instead of throwing, why not simply log an error. We've followed this patter previously (file_talker, etc).

Reply:
I'll just assert as @thomas-moulard had suggested in https://github.com/ros-tooling/system_metrics_collector/pull/16.

The reason for that is we don't want `setupStart()` and `setupStop()` to partially initialize/de-initialize, for example set/cancel the `measurement_timer_` but not the `publish_timer_`.

None of this could really happen anyway as `setupStart()`/`setupStop()` cannot be called by anything other than `Collector::start()`/`Collector::stop()`. Logging an error would be too silent.

Reply:
I don't agree with logic. Stop was meant to be treated as a teardown function, so throwing an exception if start was never called is undesirable. Logging in this case is appropriate.

If we want to check for this case if started, then we should add an atomic started_ member to the PeriodicMeasurementNode.

Reply:
The logic is it makes more sense for log message to be in `Collector::stop()`, because that's something that the user can actually call. `PeriodicMeasurementNode::setupStop()` is already guarded by the logic in `Collector::stop()` from being called if the node has not already started.

Reply:
> The logic is it makes more sense for log message to be in `Collector::stop()`, because that's something that the user can actually call. `PeriodicMeasurementNode::setupStop()` is already guarded by the logic in `Collector::stop()` from being called if the node has not already started.

Then we need the relevant checks there as well, otherwise setupStop will throw in an "ok" case.

> If `PeriodicMeasurementNode::setupStop()` were called incorrectly, that implies a logic error or bug in our code, which warrants more than a log message.

Again, I don't agree - there's no logic error in attempting to tear down resources that weren't created. If resources were started / created in the constructor then I would agree.

Re-reading @thomas-moulard 's post above, to me he means throw if stop was already called. This would warrant the atomic state variable I mentioned previously. My fear is without using the state variable the handling of state based solely on member initialization state is complicated, dangerous, and error prone.

Reply:
`setupStop()` is not supposed to be called if `setupStart()` has not been previously called. Full stop. Regardless if resources are going to be torn down or not.

Reply:
And the only way `setupStop()` is going to be called is via the implementation in `Collector`. And `Collector` does have this "atomic" state variable that you want (well, it's a `bool` and `mutex` pair, as opposed to a `atomic<bool>`).

Reply:
Yup, I'm talking with @thomas-moulard and had forgot about that. So if we assert in setupStart / setupStop with the correct state then it's fine to throw.

Reply:
So what more changes are needed?

Reply:
> So what more changes are needed?

Since setupStart is only called if the Collector has been started, we can assert that the publishers, timers, etc are non-null while throwing otherwise. 

Does this make sense? It's the conclusion that @thomas-moulard and I came to during a hallway discussion. If it isn't clear let's chat on a call, as this is a fairly lengthy thread (my bad).

Reply:
Do you mean `assert(...)` or `if (...) throw ...`? It doesn't make sense to have both.

This PR is already doing `assert(...)`, and I think it's fine.